onnxruntime/core/session/inference_session.h

```diff
@@ -733,6 +733,41 @@ class InferenceSession {
   // The life-cycle of the cache itself is maintained by the user and the user will ensure
   // the cache is valid until any session reliant on it is still in scope.
   PrepackedWeightsContainer* prepacked_weights_container_ = nullptr;
+
+  // Cache the EP instance if the user has configured the EP to capture a graph
+  // for the model and all the necessary criteria for graph capture has been met.
+  // At Run() time, if this member is not nullptr and the captured graph is ready
+  // to replay, simply invoke ReplayGraph().
+  struct CachedExecutionProviderForGraphReplay {
+    CachedExecutionProviderForGraphReplay() {}
+
+    CachedExecutionProviderForGraphReplay(IExecutionProvider* execution_provider) {
+      cached_execution_provider_for_graph_replay_ = execution_provider;
+    }
+
+    void SetExecutionProvider(IExecutionProvider* execution_provider) {
+      cached_execution_provider_for_graph_replay_ = execution_provider;
+    }
+
+    bool IsGraphCaptureEnabled() const {
+      return cached_execution_provider_for_graph_replay_ != nullptr && cached_execution_provider_for_graph_replay_->IsGraphCaptureEnabled();
+    }
+
+    bool IsGraphCaptured() const {
+      return cached_execution_provider_for_graph_replay_ != nullptr && cached_execution_provider_for_graph_replay_->IsGraphCaptured();
+    }
+
+    Status ReplayGraph() {
+      if (cached_execution_provider_for_graph_replay_) {
```

Comment:
It'll be good to check for IsGraphCaptured here as well.

Reply:
`ORT_ENFORCE(IsGraphCaptured());` is added.

Reply:
Can you please replace ORT_ENFORCE with returning Status? ORT_ENFORCE is generally used in exceptional conditions. I believe the user can correct the error if an appropriate error is returned.

Reply:
Doesn't the EP's ReplayGraph() already enforce that ? This seems redundant.